topdown/builtins/builtins.go

```diff
@@ -28,6 +29,76 @@ func (c Cache) Get(k interface{}) (interface{}, bool) {
 	return v, ok
 }
 
+type NDBCache map[string]map[ast.Value]ast.Value
+
+// Put updates the cache for the named built-in.
+// Automatically creates the 2-level hierarchy as needed.
+func (c NDBCache) Put(name string, k, v ast.Value) {
+	if m, ok := c[name]; ok {
+		m[k] = v
+	} else {
+		c[name] = map[ast.Value]ast.Value{}
+		c[name][k] = v
+	}
+}
+
+// Get returns the cached value for k for the named builtin.
+func (c NDBCache) Get(name string, k ast.Value) (ast.Value, bool) {
+	if m, ok := c[name]; ok {
+		v, ok := m[k]
+		return v, ok
+	}
+	return nil, false
+}
+
+// Convenience functions for serializing the data structure.
+func (c NDBCache) MarshalJSON() ([]byte, error) {
+	out := make(map[string]json.RawMessage)
+	for bname, m := range c {
+		// Because nested maps are a pain to serialize, we convert the lower
+		// map into an ast.Object, and serialize that.
+		entries := ast.NewObject()
+		for k, v := range m {
+			entries.Insert(ast.NewTerm(k), ast.NewTerm(v))
+		}
+		ej, err := json.Marshal(ast.NewTerm(entries))
```

Comment:
At first, I wanted to generate just a nested object JSON structure, but Golang makes that surprisingly difficult to accomplish. After a bit of flailing around, I gave up, and decided to simply convert the key-value storage parts of the cache into `ast.Object`s, since serialization/deserialization of those is something we have existing tooling for.

I tried a few approaches around serializing just the raw `ast.Object` value, but unfortunately, deserialization of the default rendering of an Object is tricky (it looks like an array of arrays of objects). I chose instead to wrapper the Object up in a Term, since that is trivial to serialize and then deserialize later (it annotates the type information needed at the top level).